### ISSUE TITLE
Dask workers don't reload modified modules

### ISSUE THREAD

**Issue by alexsbromberg** [dask/distributed#4245]
I'm trying to test a simple example of uploading a local module to a dask distributed cluster. This is intended to test a larger workflow where i'd make edits to a local module and then want to rerun those edits on the local cluster.

I have a directory structure that looks like
```python
main.py
custom_pkg/
    compute.py
    load.py
```

I expect to run `main.py` locally and upload `custom_pkg` to my dask workers.

I'm running two dask workers and the scheduler on my local machine using the Dockerfiles defined here (https://github.com/dask/dask-docker/). My only modification is that i'm using python3.7 instead of python3.8

**What happened**:
The first time i run `main.py` i get my expected results. compute_pkg.zip is uploaded and the task is able to run with that code. I then make a modification to `compute.py` and rerun `main.py`, but i get the same output as my first run.

**What you expected to happen**:
I expect client.upload_file to reload the module on the workers so that the second time I run `main.py` i get different output that reflects my changes to compute.py

**Minimal Complete Verifiable Example**:

main.py
```python
import os
import zipfile
import time 

import dask
from dask.distributed import Client

from custom_pkg.load import load_df_from_redis
from custom_pkg.compute import work_function

client = Client('localhost:8786')

def zipdir(path, ziph):
    # ziph is zipfile handle
    for root, _, files in os.walk(path):
        for file in files:
            ziph.write(os.path.join(root, file))


def create_archive():
    zipf = zipfile.ZipFile('custom_pkg.zip', 'w', zipfile.ZIP_DEFLATED)
    zipdir('custom_pkg/', zipf)
    zipf.close()


def run_computation(dist_key):
    data = dask.delayed(load_df_from_redis)(dist_key)
    result = dask.delayed(work_function)(data)
    return result

output = []
for i in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']:
    result = run_computation(i)
    output.append(result)

create_archive()
client.upload_file("custom_pkg.zip")
output = dask.compute(*output)

print(output)
```

custom_pkg/load.py
```python
def load_df_from_redis(dist_key):
    """
    dist_key 'a' to 'j' correspond to numbers 1 to 10
    """

    r = {"a": 1, "b": 2, "c": 3, "d": 4, "e": 5, "f": 6, "g": 7, "h": 8, "i": 9, "j" 10}
    val = int(r.get(dist_key))
    return val
```

custom_pkg/compute.py
```python
def work_function(data):
    # some some work on some raw data
    result = data * 20 + 53
    return result
```
**Anything else we need to know?**:
This is my first time using dask, maybe i'm missing some obvious step here? I did confirm that the new `custom_pkg.zip` is in fact uploaded to the worker, it just appears that it's not being reloaded into the python process.

**Environment**:

- Dask version: 2.30
- Python version: 3.7
- Operating System: ubuntu
- Install method (conda, pip, source): docker

**Comment by jrbourbeau** [comment]
Thanks for raising an issue @alexsbromberg! Are you able to provide a minimal reproducible example (https://blog.dask.org/2018/02/28/minimal-bug-reports)? 

Note that if the task you're submitting when you're re-running the computation has the same key as the original task (before re-uploading), and is still stored on the cluster, then the dask won't attempt to recompute and instead will just use the previous result. If using `Client.submit` you might consider setting the `pure=` keyword argument to `False`. For example here's one of our tests that does this: 

https://github.com/dask/distributed/blob/0f4c1dabb23016db14966929406138db8ee29cf8/distributed/tests/test_client.py#L1522-L1536

**Comment by alexsbromberg** [comment]
thanks @jrbourbeau I removed the redis junk from example, so it should be reproducable on any dask cluster. Is there anything else you think i'm missing?

I tried setting `pure=False`, but that didn't have an effect. What's making me think the issue is with upload_file is that if i put everything into one module like this an run it, I get different results (as expected) whenever i update the `work_function` function.

```python
import dask
from dask.distributed import Client

client = Client('localhost:8786')

def load_df_from_redis(dist_key):
    r = {"a": 1, "b": 2, "c": 3, "d": 4, "e": 5, "f": 6, "g": 7, "h": 8, "i": 9, "j" 10}
    val = int(r.get(dist_key))
    return val

def work_function(data):
    # some some work on some raw data
    result = data * 20 + 53
    return result

def run_computation(dist_key):
    data = dask.delayed(load_df_from_redis)(dist_key)
    result = dask.delayed(work_function)(data)
    return result

output = []
for i in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']:
    result = run_computation(i)
    output.append(result)

output = dask.compute(*output)

print(output)
```

**Comment by jrbourbeau** [comment]
> Is there anything else you think i'm missing?

I don't see any `Client.upload_file` calls

Thanks for updating the example. When I take your example, move `work_function` to a separate Python module, and add `client.upload_file(<module-with-work_function>)` I find that when I update the file with `work_function` and rerun the computation that the final result changes as expected

**Comment by alexsbromberg** [comment]
doh, i added the call, not sure how that got cut out.

When you call `client.upload_file` are you calling it on a single `.py` function, or on a `.zip` archive like I am? (not sure if that makes a difference.)

**Comment by alexsbromberg** [comment]
following up, this does seem to only affect me when i upload a `.zip`. If i upload a single module, change it, and repload, things work as expected.  This is the single module example I used:

main.py
```python
import dask
from dask.distributed import Client

import custom_pkg
from myfile import load_df_from_redis, work_function

client = Client('localhost:8786')


def run_computation(dist_key):
    data = dask.delayed(load_df_from_redis)(dist_key)
    result = dask.delayed(work_function)(data)
    return result

output = []
for i in ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']:
    result = run_computation(i)
    output.append(result)

def clear_cache(dask_worker):
    from distributed.worker import cache_loads
    cache_loads.data.clear()
    print("cleared data")


result = client.run(clear_cache)
client.upload_file("myfile.py")

output = dask.compute(*output)

print(output)
```

myfile.py
```python
def load_df_from_redis(dist_key):
    r = {"a": 1, "b": 2, "c": 3, "d": 4, "e": 5, "f": 6, "g": 7, "h": 8, "i": 9, "j" 10}
    val = int(r.get(dist_key))
    return val

def work_function(data):
    # some some work on some raw data
    result = data * 20 + 53
    return result
```

**Comment by jsignell** [comment]
I'm going to close this issue since it seems like there is a path forward, but please reopen if there is something that I am missing.

**Comment by Danferno** [comment]
Adding a comment here because a small detail can cause this to still go wrong (and this is the thread you find when googling the issue). 

In the current syntax you would normally use `client.register_plugin(UploadDirectory(utilsPath, update_path=True))` rather than `upload_file` and then use `client.restart()` to restart the workers. As long as your changes happen *within* the function, that work fine, as the workers will then use the updated function `utilspath`. However, I kept getting errors when I would add new functions to one of the modules, or changed the function names.

The issue is that the code above updates the workers, but not the scheduler! I wrote a helper function that reloads only the relevant functions, in my case they all start with 'utils'.
```
    def alignScheduler():
        import sys
        from importlib import reload
        utils_modules = [module for moduleName, module in sys.modules.items() if moduleName.startswith('utils')]
        for module in utils_modules:
            reload(module)
        client.run_on_scheduler(alignScheduler)
```
It might work to just reload all modules, but that's [likely to lead to unexpected bugs](https://stackoverflow.com/questions/45405600/how-to-reload-all-imported-modules). Instead, I'd recommend replacing the `moduleName.startswith()` part with a filtering function applicable to your situation.
